cpp/src/arrow/python/numpy_convert.cc

```diff
@@ -274,5 +275,348 @@ Status TensorToNdarray(const std::shared_ptr<Tensor>& tensor, PyObject* base,
+  PyAcquireGIL lock;
+
+  int type_num_data;
+  RETURN_NOT_OK(GetNumPyType(*sparse_tensor->type(), &type_num_data));
+  PyArray_Descr* dtype_data = PyArray_DescrNewFromType(type_num_data);
+  RETURN_IF_PYERROR();
+
+  const auto& sparse_index = arrow::internal::checked_cast<const SparseCSRIndex&>(
+      *sparse_tensor->sparse_index());
+  const std::shared_ptr<arrow::NumericTensor<arrow::Int64Type>> sparse_index_indptr =
+      sparse_index.indptr();
+  const std::shared_ptr<arrow::NumericTensor<arrow::Int64Type>> sparse_index_indices =
+      sparse_index.indices();
+
+  std::vector<npy_intp> npy_shape_data({sparse_index.non_zero_length(), 1});
+
+  const int ndim_indptr = sparse_index_indptr->ndim();
+  std::vector<npy_intp> npy_shape_indptr(ndim_indptr);
+
+  for (int i = 0; i < ndim_indptr; ++i) {
+    npy_shape_indptr[i] = sparse_index_indptr->shape()[i];
+  }
+
+  const int ndim_indices = sparse_index_indices->ndim();
+  std::vector<npy_intp> npy_shape_indices(ndim_indices);
+
+  for (int i = 0; i < ndim_indices; ++i) {
+    npy_shape_indices[i] = sparse_index_indices->shape()[i];
+  }
+
+  const void* immutable_data = nullptr;
+  if (sparse_tensor->data()) {
+    immutable_data = sparse_tensor->data()->data();
+  }
+
+  const void* immutable_indptr = nullptr;
+  if (sparse_index_indptr->data()) {
+    immutable_indptr = sparse_index_indptr->data()->data();
+  }
+
+  const void* immutable_indices = nullptr;
+  if (sparse_index_indices->data()) {
+    immutable_indices = sparse_index_indices->data()->data();
+  }
+
+  // Remove const =(
+  void* mutable_data = const_cast<void*>(immutable_data);
+  void* mutable_indptr = const_cast<void*>(immutable_indptr);
+  void* mutable_indices = const_cast<void*>(immutable_indices);
+
+  int array_flags = 0;
+  if (sparse_tensor->is_mutable()) {
+    array_flags |= NPY_ARRAY_WRITEABLE;
+  }
+
+  int type_num_indptr;
+  RETURN_NOT_OK(GetNumPyType(*sparse_index_indptr->type(), &type_num_indptr));
+  PyArray_Descr* dtype_indptr = PyArray_DescrNewFromType(type_num_indptr);
+
+  int type_num_indices;
+  RETURN_NOT_OK(GetNumPyType(*sparse_index_indptr->type(), &type_num_indices));
+  PyArray_Descr* dtype_indices = PyArray_DescrNewFromType(type_num_indices);
+
+  PyObject* result_data =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_data, 1, npy_shape_data.data(), nullptr,
+                           mutable_data, array_flags, nullptr);
+  PyObject* result_indptr =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_indptr, 1, npy_shape_indptr.data(),
+                           nullptr, mutable_indptr, array_flags, nullptr);
+  PyObject* result_indices =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_indices, 1, npy_shape_indices.data(),
+                           nullptr, mutable_indices, array_flags, nullptr);
+  RETURN_IF_PYERROR()
+
+  if (base == Py_None || base == nullptr) {
+    base = py::wrap_sparse_tensor_csr(sparse_tensor);
+  } else {
+    Py_XINCREF(base);
```

Comment:
You need to INCREF this 3 times as you're calling `PyArray_SetBaseObject` 3 times below.